File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleViewHandler.java
package net.gnehzr.tnoodle.server.webscrambles;

import static net.gnehzr.tnoodle.utils.GwtSafeUtils.azzert;
import static net.gnehzr.tnoodle.utils.GwtSafeUtils.parseExtension;
import static net.gnehzr.tnoodle.utils.GsonUtils.GSON;
import static net.gnehzr.tnoodle.utils.Utils.throwableToString;

import net.gnehzr.tnoodle.svglite.Svg;
import net.gnehzr.tnoodle.utils.GsonUtils;
import net.gnehzr.tnoodle.utils.BadLazyClassDescriptionException;
import net.gnehzr.tnoodle.utils.LazyInstantiatorException;
import java.lang.reflect.Type;

import com.google.gson.JsonElement;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;

import net.gnehzr.tnoodle.svglite.Color;
import net.gnehzr.tnoodle.svglite.Dimension;
import java.awt.image.BufferedImage;
import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.InvocationTargetException;
import java.util.Date;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.SortedMap;
import java.util.logging.Level;
import java.util.logging.Logger;

import javax.imageio.ImageIO;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.apache.batik.dom.svg.SVGDOMImplementation;
import org.apache.batik.util.SVGConstants;
import org.apache.batik.transcoder.TranscoderInput;
import org.apache.batik.transcoder.TranscoderOutput;
import org.apache.batik.transcoder.TranscoderException;
import org.apache.batik.transcoder.TranscodingHints;
import org.apache.batik.transcoder.image.ImageTranscoder;

import net.gnehzr.tnoodle.scrambles.InvalidScrambleException;
import net.gnehzr.tnoodle.scrambles.Puzzle;
import net.gnehzr.tnoodle.scrambles.PuzzleIcon;
import net.gnehzr.tnoodle.scrambles.PuzzlePlugins;
import net.gnehzr.tnoodle.scrambles.PuzzleImageInfo;
import net.gnehzr.tnoodle.server.SafeHttpServlet;
import net.gnehzr.tnoodle.utils.LazyInstantiator;
import net.lingala.zip4j.exception.ZipException;

import org.w3c.dom.DOMImplementation;

import com.itextpdf.text.DocumentException;

@SuppressWarnings("serial")
public class ScrambleViewHandler extends SafeHttpServlet {
    private static final Logger l = Logger.getLogger(ScrambleViewHandler.class.getName());

    private SortedMap<String, LazyInstantiator<Puzzle>> scramblers;

    public ScrambleViewHandler() throws IOException, BadLazyClassDescriptionException {
        this.scramblers = PuzzlePlugins.getScramblers();
    }


    // Copied from http://bbgen.net/blog/2011/06/java-svg-to-bufferedimage/
    class BufferedImageTranscoder extends ImageTranscoder {
        @Override
        public BufferedImage createImage(int w, int h) {
            BufferedImage bi = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
            return bi;
        }

        @Override
        public void writeImage(BufferedImage img, TranscoderOutput output) {
            this.img = img;
        }

        public BufferedImage getBufferedImage() {
            return img;
        }
        private BufferedImage img = null;
    }

    static {
        GsonUtils.registerTypeHierarchyAdapter(Puzzle.class, new Puzzlerizer());
    }

    private static class Puzzlerizer implements JsonSerializer<Puzzle>, JsonDeserializer<Puzzle> {
        @Override
        public Puzzle deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            try {
                String scramblerName = json.getAsString();
                SortedMap<String, LazyInstantiator<Puzzle>> scramblers = PuzzlePlugins.getScramblers();
                LazyInstantiator<Puzzle> lazyScrambler = scramblers.get(scramblerName);
                if(lazyScrambler == null) {
                    throw new JsonParseException(scramblerName + " not found in: " + scramblers.keySet());
                }
                return lazyScrambler.cachedInstance();
            } catch(Exception e) {
                throw new JsonParseException(e);
            }
        }

        @Override
        public JsonElement serialize(Puzzle scrambler, Type typeOfT, JsonSerializationContext context) {
            return new JsonPrimitive(scrambler.getShortName());
        }
    }

    @Override
    protected void wrappedService(HttpServletRequest request, HttpServletResponse response, String[] path, LinkedHashMap<String, String> query) throws ServletException, IOException, IllegalArgumentException, SecurityException, InstantiationException, IllegalAccessException, InvocationTargetException, ClassNotFoundException, NoSuchMethodException, DocumentException, ZipException, BadLazyClassDescriptionException, LazyInstantiatorException {
        if (path.length == 0) {
            sendError(request, response, "Please specify a puzzle.");
            return;
        }
        String[] name_extension = parseExtension(path[0]);
        String name = name_extension[0];
        String extension = name_extension[1];
        if (extension == null) {
            sendError(request, response, "Please specify an extension");
            return;
        }

        if (extension.equals("png") || extension.equals("json") || extension.equals("svg")) {
            String puzzle = name;
            LazyInstantiator<Puzzle> lazyScrambler = scramblers.get(puzzle);
            if (lazyScrambler == null) {
                sendError(request, response, "Invalid scrambler: " + puzzle);
                return;
            }
            Puzzle scrambler = lazyScrambler.cachedInstance();
            HashMap<String, Color> colorScheme = scrambler
                    .parseColorScheme(query.get("scheme"));
            String scramble = query.get("scramble");

            if (extension.equals("png")) {
                ByteArrayOutputStream bytes;
                if (query.containsKey("icon")) {
                    bytes = PuzzleIcon.loadPuzzleIconPng(scrambler.getShortName());
                } else {
                    Svg svg;
                    try {
                        svg = scrambler.drawScramble(scramble, colorScheme);
                    } catch(InvalidScrambleException e) {
                        sendText(request, response, throwableToString(e));
                        return;
                    }
                    ByteArrayInputStream svgFile = new ByteArrayInputStream(svg.toString().getBytes());

                    Dimension size = scrambler.getPreferredSize();
                    BufferedImageTranscoder imageTranscoder = new BufferedImageTranscoder();

                    // Copied from http://stackoverflow.com/a/6634963
                    // with some tweaks.
                    DOMImplementation impl = SVGDOMImplementation.getDOMImplementation();
                    TranscodingHints hints = new TranscodingHints();
                    hints.put(ImageTranscoder.KEY_WIDTH, Float.valueOf(size.width));
                    hints.put(ImageTranscoder.KEY_HEIGHT, Float.valueOf(size.height));
                    hints.put(ImageTranscoder.KEY_DOM_IMPLEMENTATION, impl);
                    hints.put(ImageTranscoder.KEY_DOCUMENT_ELEMENT_NAMESPACE_URI,SVGConstants.SVG_NAMESPACE_URI);
                    hints.put(ImageTranscoder.KEY_DOCUMENT_ELEMENT_NAMESPACE_URI,SVGConstants.SVG_NAMESPACE_URI);
                    hints.put(ImageTranscoder.KEY_DOCUMENT_ELEMENT, SVGConstants.SVG_SVG_TAG);
                    hints.put(ImageTranscoder.KEY_XML_PARSER_VALIDATING, false);

                    imageTranscoder.setTranscodingHints(hints);

                    TranscoderInput input = new TranscoderInput(svgFile);
                    try {
                        imageTranscoder.transcode(input, null);
                    } catch(TranscoderException e) {
                        l.log(Level.SEVERE, "Failed to rasterize SVG.", e);
                        sendText(request, response, throwableToString(e));
                        return;
                    }

                    BufferedImage img = imageTranscoder.getBufferedImage();
                    bytes = new ByteArrayOutputStream();
                    ImageIO.write(img, "png", bytes);
                }

                response.setHeader("Content-Type", "image/png");
                response.setContentLength(bytes.size());
                bytes.writeTo(response.getOutputStream());
            } else if(extension.equals("svg")) {
                Svg svg;
                try {
                    svg = scrambler.drawScramble(scramble, colorScheme);
                } catch(InvalidScrambleException e) {
                    sendText(request, response, throwableToString(e));
                    return;
                }

                response.setHeader("Content-Type", "image/svg+xml");
                byte[] bytes = svg.toString().getBytes();
                response.setContentLength(bytes.length);
                response.getOutputStream().write(bytes, 0, bytes.length);
            } else if (extension.equals("json")) {
                sendJSON(request, response, GSON.toJson(new PuzzleImageInfo(scrambler)));
            } else {
                azzert(false);
            }
        } else if (extension.equals("pdf") || extension.equals("zip")) {
            if (!request.getMethod().equals("POST")) {
                sendText(request, response, "You must POST to this url. Copying and pasting the url won't work.");
                return;
            }

            BufferedReader in = new BufferedReader(new InputStreamReader(request.getInputStream()));
            StringBuilder body = new StringBuilder();
            String line;
            while ((line = in.readLine()) != null) {
                body.append(line);
            }
            query = parseQuery(body.toString());

            String json = query.get("sheets"); // all the scrambles
            System.out.println(json);
            ScrambleRequest[] scrambleRequests = GSON.fromJson(json, ScrambleRequest[].class);

            String password = query.get("password");
            String generationUrl = query.get("generationUrl");
            
            Date generationDate = new Date();
            String globalTitle = name;

            if (extension.equals("pdf")) {
                ByteArrayOutputStream totalPdfOutput = ScrambleRequest
                        .requestsToPdf(globalTitle, generationDate, scrambleRequests, password);
                response.setHeader("Content-Disposition", "inline");

                // Workaround for Chrome bug with saving PDFs:
                //  https://bugs.chromium.org/p/chromium/issues/detail?id=69677#c35
                response.setHeader("Cache-Control", "public");

                sendBytes(request, response, totalPdfOutput, "application/pdf");
            } else if (extension.equals("zip")) {
                String schedule = query.remove("schedule");
                
                ByteArrayOutputStream zipOutput = ScrambleRequest
                        .requestsToZip(getServletContext(), globalTitle, generationDate, scrambleRequests, password, generationUrl, schedule, json);

                String safeTitle = globalTitle.replaceAll("\"", "'");
                response.setHeader("Content-Disposition", "attachment; filename=\"" + safeTitle + ".zip\"");
                sendBytes(request, response, zipOutput, "application/zip");
            } else {
                azzert(false);
            }
        } else {
            sendError(request, response, "Invalid extension: " + extension);
        }
    }

    static {
        GsonUtils.registerTypeAdapter(PuzzleImageInfo.class, new PuzzleImageInfoizer());
    }
    private static class PuzzleImageInfoizer implements JsonSerializer<PuzzleImageInfo> {
        @Override
        public JsonElement serialize(PuzzleImageInfo pii, Type typeOfT, JsonSerializationContext context) {

            return context.serialize(pii.toJsonable());
        }
    }

}

File: webscrambles/src/main/kotlin/org/worldcubeassociation/tnoodle/server/webscrambles/ScrambleViewHandler.kt
package org.worldcubeassociation.tnoodle.server.webscrambles

import com.google.gson.*
import io.ktor.application.call
import io.ktor.http.ContentType
import io.ktor.request.receiveText
import io.ktor.response.header
import io.ktor.response.respond
import io.ktor.response.respondBytes
import io.ktor.response.respondText
import io.ktor.routing.Routing
import io.ktor.routing.get
import io.ktor.routing.post
import io.ktor.util.toMap
import io.netty.handler.codec.http.QueryStringDecoder
import net.gnehzr.tnoodle.scrambles.*
import org.apache.batik.anim.dom.SVGDOMImplementation
import org.apache.batik.transcoder.TranscoderInput
import org.apache.batik.transcoder.TranscoderOutput
import org.apache.batik.transcoder.TranscodingHints
import org.apache.batik.transcoder.image.ImageTranscoder
import org.apache.batik.util.SVGConstants
import org.worldcubeassociation.tnoodle.server.RouteHandler
import net.gnehzr.tnoodle.scrambles.PuzzleImageInfo.infoToJsonable

import javax.imageio.ImageIO
import java.awt.image.BufferedImage
import java.io.*
import java.lang.reflect.Type
import java.util.Date

object ScrambleViewHandler : RouteHandler {
    private val scramblers = PuzzlePlugins.getScramblers()

    // Copied from http://bbgen.net/blog/2011/06/java-svg-to-bufferedimage/
    internal class BufferedImageTranscoder : ImageTranscoder() {
        var bufferedImage: BufferedImage? = null
            private set

        override fun createImage(w: Int, h: Int): BufferedImage {
            return BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB)
        }

        override fun writeImage(img: BufferedImage, output: TranscoderOutput) {
            this.bufferedImage = img
        }
    }

    override fun install(router: Routing) {
        router.get("/view/{puzzleExt?}") {
            val puzzleExt = call.parameters["puzzleExt"]

            if (puzzleExt == null) {
                call.respondText("Please specify a puzzle")
                return@get
            }

            val (name, extension) = puzzleExt.split(".", limit = 2)
            val scrambler = scramblers[name]

            if (scrambler == null) {
                call.respondText("Invalid scrambler: $name")
                return@get
            }

            val query = call.request.queryParameters.toMap().mapValues { it.value.first() }

            val colorScheme = query["scheme"]?.let { scrambler.parseColorScheme(it) } ?: mapOf()
            val scramble = query["scramble"]

            when (extension) {
                "png" -> {
                    if (query.containsKey("icon")) {
                        val icon = PuzzleIcon.loadPuzzleIconPng(scrambler.shortName)

                        call.respondBytes(icon, ContentType.Image.PNG)
                    } else {
                        val svg = scrambler.drawScramble(scramble, colorScheme)
                        val svgFile = svg.toString().byteInputStream()

                        val (width, height) = scrambler.preferredSize
                        val imageTranscoder = BufferedImageTranscoder()

                        // Copied from http://stackoverflow.com/a/6634963
                        // with some tweaks.
                        val impl = SVGDOMImplementation.getDOMImplementation()

                        val hints = TranscodingHints().apply {
                            this[ImageTranscoder.KEY_WIDTH] = width.toFloat()
                            this[ImageTranscoder.KEY_HEIGHT] = height.toFloat()
                            this[ImageTranscoder.KEY_DOM_IMPLEMENTATION] = impl
                            this[ImageTranscoder.KEY_DOCUMENT_ELEMENT_NAMESPACE_URI] = SVGConstants.SVG_NAMESPACE_URI
                            this[ImageTranscoder.KEY_DOCUMENT_ELEMENT_NAMESPACE_URI] = SVGConstants.SVG_NAMESPACE_URI
                            this[ImageTranscoder.KEY_DOCUMENT_ELEMENT] = SVGConstants.SVG_SVG_TAG
                            this[ImageTranscoder.KEY_XML_PARSER_VALIDATING] = false
                        }

                        imageTranscoder.transcodingHints = hints

                        val input = TranscoderInput(svgFile)
                        imageTranscoder.transcode(input, null)

                        val img = imageTranscoder.bufferedImage

                        val bytes = ByteArrayOutputStream().also { ImageIO.write(img, "png", it) }

                        call.respondBytes(bytes.toByteArray(), ContentType.Image.PNG)
                    }
                }
                "svg" -> {
                    val svg = scrambler.drawScramble(scramble, colorScheme)

                    call.respondText(svg.toString(), ContentType.Image.SVG)
                }
                "json" -> call.respond(scrambler.infoToJsonable())
                else -> call.respondText("Invalid extension: $extension")
            }
        }

        router.post("/view/{puzzleExt?}") {
            val puzzleExt = call.parameters["puzzleExt"]

            if (puzzleExt == null) {
                call.respondText("Please specify a puzzle")
                return@post
            }

            val (name, extension) = puzzleExt.split(".", limit = 2)
            val scrambler = scramblers[name]

            if (scrambler == null) {
                call.respondText("Invalid scrambler: $name")
                return@post
            }

            if (extension == "pdf" || extension == "zip") {
                val body = call.receiveText()
                val query = QueryStringDecoder(body).parameters().mapValues { it.value.first() }

                val scrambleRequests = GSON.fromJson(query["sheets"], Array<ScrambleRequest>::class.java)
                val password = query["password"]

                val generationDate = Date()

                when (extension) {
                    "pdf" -> {
                        val totalPdfOutput = ScrambleRequest
                            .requestsToPdf(name, generationDate, scrambleRequests, password)

                        call.response.header("Content-Disposition", "inline")

                        // Workaround for Chrome bug with saving PDFs:
                        // https://bugs.chromium.org/p/chromium/issues/detail?id=69677#c35
                        call.response.header("Cache-Control", "public")

                        call.respondBytes(totalPdfOutput.toByteArray(), ContentType.Application.Pdf)
                    }
                    "zip" -> {
                        val generationUrl = query["generationUrl"]
                        val schedule = query["schedule"]

                        val wcifHelper = WCIFHelper(schedule, scrambleRequests)

                        val zipOutput = ScrambleRequest
                            .requestsToZip(getServletContext(), name, generationDate, scrambleRequests, password, generationUrl, wcifHelper)

                        val safeTitle = name.replace("\"".toRegex(), "'")

                        call.response.header("Content-Disposition", "attachment; filename=\"$safeTitle.zip\"")
                        call.respondBytes(zipOutput.toByteArray(), ContentType.Application.Zip)
                    }
                }
            } else {
                call.respondText("Invalid extension: $extension")
            }
        }
    }

    private class Puzzlerizer : JsonSerializer<Puzzle>, JsonDeserializer<Puzzle> {
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Puzzle {
            try {
                val scramblerName = json.asString
                val scramblers = PuzzlePlugins.getScramblers()

                return scramblers[scramblerName]
                    ?: throw JsonParseException(scramblerName + " not found in: " + scramblers.keys)
            } catch (e: Exception) {
                throw JsonParseException(e)
            }
        }

        override fun serialize(scrambler: Puzzle, typeOfT: Type, context: JsonSerializationContext): JsonElement {
            return JsonPrimitive(scrambler.shortName)
        }
    }

    private class PuzzleImageInfoizer : JsonSerializer<PuzzleImageInfo> {
        override fun serialize(pii: PuzzleImageInfo, typeOfT: Type, context: JsonSerializationContext): JsonElement {
            return context.serialize(pii.toJsonable())
        }
    }

    init {
        GsonUtils.registerTypeHierarchyAdapter(Puzzle::class.java, Puzzlerizer())
        GsonUtils.registerTypeAdapter(PuzzleImageInfo::class.java, PuzzleImageInfoizer())
    }
}
